cmake/FindOpenCL.cmake

```diff
@@ -1,65 +1,143 @@
+# 
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+# 
+################################################################################
+
+include( FindPackageHandleStandardArgs )
+find_package_handle_standard_args(
+    OpenCL
+    FOUND_VAR OpenCL_FOUND
+    REQUIRED_VARS
+        OpenCL_LIBRARIES
+        OpenCL_INCLUDE_DIRS
+        CL_TARGET_OpenCL_VERSION
+    VERSION_VAR OpenCL_VERSION
+)
 
-if("${CMAKE_SIZEOF_VOID_P}" EQUAL "8")
-    find_library( OPENCL_LIBRARIES
-        NAMES OpenCL
+if(OpenCL_LIBRARIES AND OpenCL_INCLUDE_DIRS)
+    set(OpenCL_FOUND TRUE)
+    add_definitions(-DCL_TARGET_OPENCL_VERSION=${CL_TARGET_OpenCL_VERSION})
+else()
+    find_path(OPENCL_INCLUDE_DIRS
+        NAMES OpenCL/cl.h CL/cl.h
         HINTS
-        ${OPENCL_ROOT}/lib
-        $ENV{AMDAPPSDKROOT}/lib
-        $ENV{CUDA_PATH}/lib
-        DOC "OpenCL dynamic library path"
-        PATH_SUFFIXES x86_64 x64 x86_64/sdk
+        $ENV{OPENCL_ROOT}/include
+        $ENV{AMDAPPSDKROOT}/include
+        $ENV{CUDA_PATH}/include
+        $ENV{OCL_ROOT}/include
         PATHS
-        /usr/lib
-        /usr/local/cuda/lib
-        /opt/cuda/lib
-        /opt/rocm/opencl/lib
-        )
-else( )
-    find_library( OPENCL_LIBRARIES
-        NAMES OpenCL
-        HINTS
-        ${OPENCL_ROOT}/lib
-        $ENV{AMDAPPSDKROOT}/lib
-        $ENV{CUDA_PATH}/lib
-        DOC "OpenCL dynamic library path"
-        PATH_SUFFIXES x86 Win32
+        ${ROCM_PATH}/include
+        /usr/include
+        /usr/local/include
+        /usr/local/cuda/include
+        /opt/cuda/include
+        DOC "OpenCL header file path"
+    )
+    mark_as_advanced( OPENCL_INCLUDE_DIRS )
 
-        PATHS
-        /usr/lib
-        /usr/local/cuda/lib
-        /opt/cuda/lib
+    if("${CMAKE_SIZEOF_VOID_P}" EQUAL "8")
+        find_library( OPENCL_LIBRARIES
+            NAMES OpenCL
+            HINTS
+            $ENV{OPENCL_ROOT}/lib
+            $ENV{AMDAPPSDKROOT}/lib
+            $ENV{CUDA_PATH}/lib
+            $ENV{OCL_ROOT}/lib
+            DOC "OpenCL dynamic library path"
+            PATH_SUFFIXES x86_64 x64 x86_64/sdk
+            PATHS
+            ${ROCM_PATH}/lib/
+            /usr/lib
+            /usr/local/cuda/lib
+            /opt/cuda/lib
+        )
+    else( )
+        find_library( OPENCL_LIBRARIES
+            NAMES OpenCL
+            HINTS
+            $ENV{OPENCL_ROOT}/lib
+            $ENV{AMDAPPSDKROOT}/lib
```

Comment:
why "APPSDK" is needed for RPP?

Reply:
This would allow us to use OpenCL modules outside ROCm. RPP OpenCL backend could then be built.